jsonrpc/debug_endpoint.go

```diff
@@ -671,6 +690,74 @@ func (d *Debug) GetRawReceipts(filter BlockNumberOrHash) (interface{}, error) {
 	)
 }
 
+func (d *Debug) TraceChain(start, end BlockNumber, config *TraceConfig) (interface{}, error) {
+	return d.throttling.AttemptRequest(
+		context.Background(),
+		func() (interface{}, error) {
+			startNum, err := GetNumericBlockNumber(start, d.store)
+			if err != nil {
+				return nil, err
+			}
+
+			endNum, err := GetNumericBlockNumber(end, d.store)
+			if err != nil {
+				return nil, err
+			}
+
+			if startNum > endNum {
+				return nil, fmt.Errorf("end block (#%d) needs to come after start block (#%d)", endNum, startNum)
+			}
+
+			blocks := int(endNum-startNum) + 1
+			results := make([]BlockTraceResult, blocks)
+			resultCh := make(chan BlockTraceResult, blocks)
+			maxConcurrency := runtime.NumCPU() - 1
+			semaphore := make(chan struct{}, maxConcurrency)
+
+			var wg sync.WaitGroup
+			for i := 0; i < blocks; i++ {
+				wg.Add(1)
+				semaphore <- struct{}{}
+
+				go func(i int) {
+					defer wg.Done()
+					defer func() { <-semaphore }()
```

Comment:
Does it make sense? If you can, just try it out, in order not to block indefinitely.

```suggestion
					defer func() {
					    wg.Done()
					    <-semaphore
					 }()
```

Also take a look to the `errgroup.Group`, maybe we can avoid having both the `semaphore` channel and `WaitGroup`:
https://pkg.go.dev/golang.org/x/sync/errgroup#Group.SetLimit